### PR TITLE
fix: an impossible mask kills the particle (weight 0) instead of aborting the SMC run

### LINKED ISSUE #47: test_haiku fails with NullMask (mask rules out all tokens) — not backend-specific
## What
`tests/test_examples.py::test_haiku[vllm]` fails under the **vLLM** backend:

```
llamppl.distributions.lmcontext.NullMask: Unable to compute log probability of mask that rules out all tokens.
```

The haiku example reaches a step where the constraint masks out *every* token, so the log-probability of the mask can't be computed.

## Context
- Surfaced once GPU CI was restored (the partner GPU runner had been auto-cancelling jobs at GitHub's 24h queue cap) and vLLM was correctly installed in the `build` job (#46).
- The other `[vllm]` tests pass: `test_examples.py::test_hard_constraints[vllm]` and `test_lmcontext.py::test_init[vllm]`. So vLLM itself works — this is specific to the haiku example under vLLM.
- Backend: `gpt2` via `genlm-backend[vllm]` (vLLM 0.10.x), `n_particles=20`.

## Current state
Scoped, non-strict **xfail** on the `[vllm]` param only (added in #46) so it doesn't block CI:
in `tests/test_examples.py::test_haiku`, `request.applymarker(pytest.mark.xfail(...))` when `backend == "vllm"`.

## Possible directions
- Could be sampling-dependent (re-run to check determinism) or a genuine edge case where the haiku syllable/structure constraint admits no continuation given gpt2's vLLM logits/tokenization.
- If real, the haiku example and/or `NullMask` handling in `llamppl/distributions/lmcontext.py` may need to handle the all-tokens-masked state gracefully.

**When fixed:** remove the xfail in `tests/test_examples.py::test_haiku`.

### PR DESCRIPTION
**Pull request by postylem**
## Problem
When a model conditions on a token mask that rules out every token (a zero-probability event), `LMTokenMask.log_prob` raised `NullMask`. Nothing catches it, so a single impossible particle aborted the entire SMC run. (The uncaught unwind could also leave a reused `LLM`'s autobatch state corrupted, surfacing later as `asyncio.InvalidStateError`.)

## Fix
Handle it the way `condition(False)` is already handled:
1. `LMTokenMask.log_prob` returns `-inf` (the correct log-probability of a zero-probability observation) instead of raising. It returns before the `model_mask` / `next_token_logprobs` updates, because applying those with an empty token set is what corrupts the context (and is what the pre-existing `-inf` code path fell through into).
2. `Model.observe` finishes the particle when an observation's log-probability is `-inf`, exactly as `condition(False)` does (`score(-inf)` then `finish()`). The impossible particle becomes weight 0 and done, and is dropped at the next resample.

Because the handling lives in `observe` and `log_prob`, it applies to every inference driver (`smc_standard`, `smc_steer`, and so on) with no per-driver changes and no new exception type. The now-unused `NullMask` exception is removed.

Closes #47.

## Deterministic reproduction
The failure is stochastic in normal use, but all sampling goes through the global numpy RNG, so seeding makes it deterministic. Save this as `repro.py` in the repo root and run it with `PYTHONPATH=. python repro.py` (needs the `hf` extra and downloads gpt2 on first run):

```python
import asyncio
import numpy as np
from llamppl import CachedCausalLM
from examples.haiku import run_example

np.random.seed(21)
LLM = CachedCausalLM.from_pretrained("gpt2", backend="hf")  # vLLM reproduces it too
asyncio.run(run_example(LLM, poem_title="The beauty of testing", n_particles=20))
```

On `main` this raises `NullMask: Unable to compute log probability of mask that rules out all tokens` and aborts the run. With this change it completes (the impossible particle is dropped at the next resample). Seeds 0 through 20 pass; seed 21 is the first that hits the dead-end.

## Test plan
Locally (hf, mock backends): `test_haiku` passes. Deterministic seed-21 completes instead of raising. GPU job on this PR checks to confirm the fix addresses what previously failed on vllm.

### CODE REVIEW

File: tests/test_examples.py [hunk 1]
```diff
@@ -44,19 +45,25 @@ def test_hard_constraints(LLM, n_particles=20, max_tokens=25):
...
     )
     assert len(particles) == n_particles
+
+
+@pytest.mark.skipif(
+    MLX_AVAILABLE,
+    reason="hf-only regression; the MLX/macOS CI job runs hf on MPS and OOMs",
+)
+def test_haiku_impossible_mask_completes_instead_of_raising():
+    # Regression for #47: with this seed the haiku example drives a particle into
```
Comment (samuki) [inline tests/test_examples.py]: Nitpick, but I found "Regression for #47:" confusing while reading the code and would just describe the problem instead.

File: tests/test_lmcontext.py [hunk 2]
```diff
@@ -62,3 +63,23 @@ async def async_context_create():
         rtol=5e-4,
         err_msg="Async context create",
     )
+
+
+def test_observe_impossible_mask_kills_particle():
+    # Regression for #47: conditioning on a mask that rules out every token is a
```
Comment (samuki) [inline tests/test_lmcontext.py]: Same as above

**Review by samuki — APPROVED** [review]

**Review by ClementeP — APPROVED** [review]
